engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/NadelCoerceTransform.kt

```diff
@@ -0,0 +1,98 @@
+package graphql.nadel.enginekt.transform
+
+import graphql.nadel.Service
+import graphql.nadel.ServiceExecutionResult
+import graphql.nadel.enginekt.NadelExecutionContext
+import graphql.nadel.enginekt.blueprint.NadelOverallExecutionBlueprint
+import graphql.nadel.enginekt.transform.NadelCoerceTransform.State
+import graphql.nadel.enginekt.transform.query.NadelQueryPath.Companion.root
+import graphql.nadel.enginekt.transform.query.NadelQueryTransformer
+import graphql.nadel.enginekt.transform.result.NadelResultInstruction
+import graphql.nadel.enginekt.transform.result.json.JsonNode
+import graphql.nadel.enginekt.transform.result.json.JsonNodeExtractor
+import graphql.nadel.enginekt.transform.result.json.JsonNodeExtractor.getNodesAt
+import graphql.nadel.enginekt.util.queryPath
+import graphql.normalized.ExecutableNormalizedField
+import graphql.schema.FieldCoordinates
+import graphql.schema.GraphQLScalarType
+import graphql.schema.GraphQLTypeUtil
+
+internal class NadelCoerceTransform : NadelTransform<State> {
+    data class State(
+        val fieldType: GraphQLScalarType,
+    )
+
+    override suspend fun isApplicable(
+        executionContext: NadelExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        services: Map<String, Service>,
+        service: Service,
+        overallField: ExecutableNormalizedField
+    ): State? {
+        val schema = executionBlueprint.schema
+
+        val types = overallField.objectTypeNames
+            .asSequence()
+            .mapNotNull {
+                schema.getFieldDefinition(
+                    FieldCoordinates.coordinates(
+                        it,
+                        overallField.fieldName
+                    )
+                )
+            }
+            .map { it.type }
+            .toList()
+
+        if (types.distinct().size != 1) {
+            // field type on all definitions should be the same
```

Comment:
maybe I should put an "assert" here to throw an error instead?

Reply:
don't think we can throw an error (at least not as it is currently), for example for a field `Foo.__typename` `disticntTypes` is 0

Reply:
also I think `distinctTypes` will be equal to 2 for some crazy edge cases like this:

```graphql
type Query {
  user: User
}
interface User {
  id: ID
  name: String
  user: User
}

type AUser implements User {
  id: ID
  name: String
  user: AUser
}

type BUser implements User {
  id: ID
  name: String
  user: BUser
}

```

Reply:
I think we need to remove the comment or update it

Reply:
Good call @temaEmelyan.  I'll add more tests for the `__typename` and the interface scenarios and update the comment 👍

Reply:
@temaEmelyan I've pushed a commit with a better comment and more documentation as well.

Reply:
very nice!